### PR TITLE
Moving scale code to web jobs sdk to resuse in SC v3.

### LINKED ISSUE #2938: Exposing common scaling code to consume by SC v3
For SC v3(redesign) we can reuse RDS code from function host such as:
1. Decision making.
`ScaleManager`
2. Metric saving. 
`ScaleMonitorService` to periodically save metrics. 
`InMemoryScaleMetricsRepository` - SC saves metrics in memory. 
3. Dynamic concurrency value reading.
`BlobStorageConcurrencyStatusRepository` - making public.

The plan is that SC v3 and function host will be referencing webjob sdk use the classes/interfaces in the PR.

### PR DESCRIPTION
**Pull request by alrod**
### Issue describing the changes in this PR

resolves https://github.com/Azure/azure-webjobs-sdk/issues/2938

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public class ScaleOptionsSetup : IConfigureOptions<ScaleOptions>
+    {
+        private readonly IOptions<FunctionsHostingConfigOptions> _functionsHostingConfigOptions;
+        private readonly IEnvironment _environment;
+
+        public ScaleOptionsSetup(IEnvironment environment, IOptions<FunctionsHostingConfigOptions> functionsHostingConfigOptions)
+        {
+            _environment = environment;
+            _functionsHostingConfigOptions = functionsHostingConfigOptions;
```
Comment (liliankasem) [inline src/WebJobs.Script/Config/ScaleOptionsSetup.cs:18]: Do we want to throw if null?

```
_environment = environment ?? throw new ArgumentNullException(nameof(environment));
_functionsHostingConfigOptions = functionsHostingConfigOptions ?? ?? throw new ArgumentNullException(nameof(functionsHostingConfigOptions));
```

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+        public void Configure(ScaleOptions options)
+        {
+            options.IsTargetBasedScalingEnabled = _environment.IsTargetBasedScalingEnabled();
+            options.IsTargetBasedScalingEnabledForTriggerFunc = IsTargetBasedScalingEnabledForTrigger;
+        }
+
+        private bool IsTargetBasedScalingEnabledForTrigger(ITargetScaler targetScaler)
+        {
+            string assemblyName = targetScaler.GetType().Assembly.GetName().Name;
```
Comment (liliankasem) [inline src/WebJobs.Script/Config/ScaleOptionsSetup.cs:29]:
```suggestion
            string assemblyName = targetScaler?.GetType().Assembly.GetName().Name;
```

File: src/WebJobs.Script/Config/ScaleOptionsSetup.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            string flag = _functionsHostingConfigOptions.Value.GetFeature(assemblyName);
+            return flag == "1";
```
Comment (liliankasem) [inline src/WebJobs.Script/Config/ScaleOptionsSetup.cs:31]: Can we make "1" a constant with a name that describes what the 1 represents

**Review by liliankasem — DISMISSED** [review]

File: test/WebJobs.Script.Tests/Scale/FunctionsScaleMonitorServiceTests.cs [hunk 4]
```diff
@@ -107,7 +105,7 @@ public async Task OnTimer_DoesNotSample_WhenNotPrimaryHost()
 
             var logs = _loggerProvider.GetAllLogMessages().ToArray();
             Assert.Equal(1, logs.Length);
-            Assert.Equal("Runtime scale monitoring is enabled.", logs[0].FormattedMessage);
+            Assert.Equal("Scale monitor service started is started.", logs[0].FormattedMessage);
```
Comment (madelinegordon) [inline test/WebJobs.Script.Tests/Scale/FunctionsScaleMonitorServiceTests.cs:108]: nit - should the message here be "Scale monitor service is started" ?

**Review by surgupta-msft — APPROVED** [review]

**Comment by liliankasem** [comment]
@alrod is this still blocked? WebJobs has been fully released

**Comment by mathewc** [comment]
> @alrod is this still blocked? WebJobs has been fully released

@liliankasem Yes this is blocked, and this PR probably should have been marked Draft since it relies on https://github.com/Azure/azure-webjobs-sdk/pull/2954.

**Comment by liliankasem** [comment]
> > @alrod is this still blocked? WebJobs has been fully released
> 
> @liliankasem Yes this is blocked, and this PR probably should have been marked Draft since it relies on [Azure/azure-webjobs-sdk#2954](https://github.com/Azure/azure-webjobs-sdk/pull/2954).

Thanks for the info! I'm going to go ahead and close this PR since it might be a while till 3.0.37 is out. Please open a new one when this is ready to go or reopen as a draft

**Pull request closed without merging**